lua/render-markdown/render/dash.lua

```diff
@@ -16,8 +16,24 @@ end
 
 function Render:render()
     local width = self.dash.width
-    width = type(width) == 'number' and width or vim.o.columns
-    local virt_text = { self.dash.icon:rep(width), self.dash.highlight }
+    local win_width = vim.api.nvim_win_get_width(0)
+    if type(width) == 'string' then
+        if width == 'full' then
+            width = win_width
+        else
+            width = width:gsub('%%', '')
+            width = tonumber(width) / 100 * win_width
+        end
+    end
+    local indent = ''
+    if self.dash.align == 'center' then
+        indent = string.rep(' ', (win_width - width) / 2)
+    elseif self.dash.align == 'right' then
+        indent = string.rep(' ', win_width - width)
+    end
+    local text = indent .. self.dash.icon:rep(width)
+
+    local virt_text = { text, self.dash.highlight }
```

Comment:
Hi, thanks for the PR!

There are a couple of components (headings & code blocks) that already support this kind of percent width / centering behavior so I'd like to keep this implementation consistent with those.

The main concepts are:

1) When floats between 0 & 1 are used for a width value it is treated as a percentage of available window width
2) Use a `left_margin` property where a value of 0.5 results in center alignment

The `context` object has a common method `resolve_offset` for computing these things based on the space already occupied. So the implementation for this would become:

```lua
local width = self.dash.width
width = type(width) == 'number' and self.context:resolve_offset(width, 0) or vim.o.columns
local margin = self.context:resolve_offset(self.dash.left_margin, width)

local virt_text = {}
if margin > 0 then
    table.insert(virt_text, self:padding_text(margin))
end
table.insert(virt_text, { self.dash.icon:rep(width), self.dash.highlight })

local start_row, end_row = self.node.start_row, self.node.end_row - 1
self.marks:add('dash', start_row, 0, {
    virt_text = virt_text,
    virt_text_pos = 'overlay',
})
if end_row > start_row then
    self.marks:add('dash', end_row, 0, {
        virt_text = virt_text,
        virt_text_pos = 'overlay',
    })
end
```

Reply:
solved